backend/infrahub/core/diff/combiner.py

```diff
@@ -1,6 +1,319 @@
-from .model.path import EnrichedDiffRoot
+from copy import deepcopy
+from dataclasses import dataclass, field
+from typing import Iterable
+from uuid import uuid4
+
+from infrahub.core.constants import DiffAction, RelationshipCardinality
+from infrahub.core.schema_manager import SchemaManager
+
+from .model.path import (
+    EnrichedDiffAttribute,
+    EnrichedDiffNode,
+    EnrichedDiffProperty,
+    EnrichedDiffRelationship,
+    EnrichedDiffRoot,
+    EnrichedDiffSingleRelationship,
+)
+
+
+@dataclass
+class NodePair:
+    earlier: EnrichedDiffNode | None = field(default=None)
+    later: EnrichedDiffNode | None = field(default=None)
 
 
 class DiffCombiner:
-    def combine(self, earlier_diff: EnrichedDiffRoot, later_diff: EnrichedDiffRoot) -> EnrichedDiffRoot:  # pylint: disable=unused-argument
-        return earlier_diff
+    def __init__(self, schema_manager: SchemaManager) -> None:
+        self.schema_manager = schema_manager
+        # {child_uuid: (parent_uuid, parent_rel_name)}
+        self._child_parent_uuid_map: dict[str, tuple[str, str]] = {}
+        self._parent_node_uuids: set[str] = set()
+        self._earlier_nodes_by_uuid: dict[str, EnrichedDiffNode] = {}
+        self._later_nodes_by_uuid: dict[str, EnrichedDiffNode] = {}
+        self._common_node_uuids: set[str] = set()
+        self._diff_branch_name: str | None = None
+
+    def _initialize(self, earlier_diff: EnrichedDiffRoot, later_diff: EnrichedDiffRoot) -> None:
+        self._diff_branch_name = earlier_diff.diff_branch_name
+        self._child_parent_uuid_map = {}
+        self._earlier_nodes_by_uuid = {}
+        self._later_nodes_by_uuid = {}
+        self._common_node_uuids = set()
+        # map the parent of each node (if it exists), preference to the later diff
+        for diff_root in (earlier_diff, later_diff):
+            for node in diff_root.nodes:
+                for rel in node.relationships:
+                    for child_node in rel.nodes:
+                        self._child_parent_uuid_map[child_node.uuid] = (node.uuid, rel.name)
+        # UUIDs of all the parents, removing the stale parents from the earlier diff
+        self._parent_node_uuids = {parent_tuple[0] for parent_tuple in self._child_parent_uuid_map.values()}
```

Comment:
It's not quite clear what `parent_tuple[0] is in this case. I can see from above that it's some form of string. Don't know if it would be easier if _child_parent_uuid_map was a dict[str, SomeObject] instead.

Reply:
I think the comment above the initialization of `self._child_parent_uuid_map` and the name of this instance variable should make it clear that `parent_tuple[0]` is the UUID of the parent node